cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -234,7 +234,7 @@ func Run(c *config.CompletedConfig, stopCh <-chan struct{}) error {
 		}
 
 		controllerContext.InformerFactory.Start(stopCh)
-		controllerContext.ObjectOrMetadataInformerFactory.Start(stopCh)
```

Comment:
does this mean that if k-c-m was already watching all pods for typed controller loops that required all pod info, this would now *additionally* watch all metadata for pods? in other words, will this double up watching both full info and metadata for some objects?

Reply:
Yes, but it should not be a big issue in reality because metadata are small so double of metadata is still smaller than having all objects in caches. The problem here -- resourcequota is a generic controller that caches all k8s objects including CRD. It's very likely to cache with excessive memory cost.

Reply:
resourcequota wouldn't watch CRDs, those are cluster-scoped. It also wouldn't watch full CR objects, since we don't have typed informers for those, so all custom resources would already have been falling back to the metadata informer.

If we already have a typed informer running for a given resource type, this change is going to be a net negative for that type. This is only beneficial for resources which we have typed informers for which would not run otherwise. Do we know what resources those are, and whether those objects are typically high volume or large in size?

Reply:
You are right about CRD and CR objects.

I was thinking ConfigMap/Secret, which are usually larger than their metadata and not watched by other controllers by default.

Reply:
the serviceaccount token controller starts an informer which watches secrets, since it has to populate them with tokens and clean up token secrets for delete serviceaccount objects:

https://github.com/kubernetes/kubernetes/blob/731238fb416a2ae73b39063acdf0c6326555a045/cmd/kube-controller-manager/app/controllermanager.go#L667

the root CA publisher controller starts an informer which watches configmaps, since it has to react to modifications or deletion of the root-ca.crt configmap in every namespace:

https://github.com/kubernetes/kubernetes/blob/731238fb416a2ae73b39063acdf0c6326555a045/cmd/kube-controller-manager/app/certificates.go#L183

Reply:
Those two controllers are both in my following commits to remove the typed informers for ConfigMap & Secret. But they also need to change other parts of code so I didn't put them in this PR to avoid making it too complicated.

For discussion, we can skip those two controllers and the other two disabled by default  https://github.com/kubernetes/kubernetes/blob/731238fb416a2ae73b39063acdf0c6326555a045/cmd/kube-controller-manager/app/controllermanager.go#L412

Reply:
this change would be less fragile and an unconditional win if it kept the fallback informer factory but adjusted its behavior from "return a typed informer falling back to a metadata informer" to "return an already-started typed informer falling back to a metadata informer"

that would let us use lightweight metadata informers when we didn't already have a running typed informer, and would mean that any follow-up cleanups to eliminate overly broad watching of typed objects would automatically let the generic informers scope down their use to metadata-only informers

I don't think we should drop the fallback and spawn parallel metadata informers for typed things we're already watching

Reply:
Changed as suggested. PTAL, thanks!